### ISSUE TITLE
Error when build manually

### ISSUE THREAD

**Issue by gogolxdong** [snej/nimdbx#9]
```shell
PS C:\Users\asus\.nimble\pkgs\nimdbx-0.4.1\libmdbx-dist> gcc -std=gnu11 -O2 -g -Wall -Werror -Wextra -Wpedantic -ffunction-sections -fPIC -fvisibility=hidden -pthread -Wno-error=attributes  -DNDEBUG=1 -DMDBX_CONFIG_H="config.h" -ULIBMDBX_EXPORTS -c mdbx.c -o mdbx-static.o
mdbx.c:16:10: error: #include expects "FILENAME" or <FILENAME>
 #include MDBX_CONFIG_H
          ^~~~~~~~~~~~~
mdbx.c:1371:29: error: unknown type name 'FILE_INFO_BY_HANDLE_CLASS'
     _In_ HANDLE hFile, _In_ FILE_INFO_BY_HANDLE_CLASS FileInformationClass,
                             ^~~~~~~~~~~~~~~~~~~~~~~~~
mdbx.c:1373:19: error: unknown type name 'MDBX_GetFileInformationByHandleEx'
 MDBX_INTERNAL_VAR MDBX_GetFileInformationByHandleEx
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mdbx.c:1392:29: error: unknown type name 'FILE_INFO_BY_HANDLE_CLASS'
     _In_ HANDLE hFile, _In_ FILE_INFO_BY_HANDLE_CLASS FileInformationClass,
                             ^~~~~~~~~~~~~~~~~~~~~~~~~
mdbx.c:1394:19: error: unknown type name 'MDBX_SetFileInformationByHandle'
 MDBX_INTERNAL_VAR MDBX_SetFileInformationByHandle
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mdbx.c: In function 'bad_page':
mdbx.c:3617:22: error: ISO C does not support the 'I64' ms_printf length modifier [-Werror=format=]
                      "corrupted page #%u, mod-txnid %" PRIaTXN " \n",
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mdbx.c: In function 'thread_key_create':
mdbx.c:4124:14: error: ISO C does not support the 'I64' ms_printf length modifier [-Werror=format=]
   mdbx_trace("&key = %p, value %" PRIuPTR ", rc %d",
...
```

**Comment by snej** [comment]
The error `#include expects "FILENAME" or <FILENAME>` implies you're building with a compiler version (or build settings) libmdbx isn't compatible with. I don't know where the `-Wpedantic` flag came from — probably Nimterop. That might be causing the error. I haven't run into this, but I build with Clang not GCC.

The errors about `FILE_INFO_BY_HANDLE_CLASS` are weird. I don't get them because they're in code that only compiles on Windows. There's no definition of that symbol anywhere, so either it's defined in some Windows header(?) or this is a bug that breaks the build of libmdbx on Windows.

It's probably best to file an issue against [libmdbx itself](https://github.com/erthink/libmdbx/issues). You can let Leonid know the `mdbx.c` file you tried to build was generated by the regular `make dist` command.

**Comment by erthink** [comment]
> ```
> cc -O2 -g -Wall -Werror -Wextra -Wpedantic -ffunction-sections -fPIC -fvisibility=hidden -std=gnu11 -pthread -Wno-error=attributes  -DNDEBUG=1 '-DMDBX_CONFIG_H="config.h"' -ULIBMDBX_EXPORTS -c mdbx.c -o mdbx-static.o
> GNUmakefile:101: recipe for target 'mdbx-static.o' failed
> process_begin: CreateProcess(NULL, cc -O2 -g -Wall -Werror -Wextra -Wpedantic -ffunction-sections -fPIC -fvisibility=hidden -std=gnu11 -pthread -Wno-error=attributes -DNDEBUG=1 -DMDBX_CONFIG_H=\"config.h\" -ULIBMDBX_EXPORTS -c mdbx.c -o mdbx-static.o, ...) failed.
> make (e=2): ϵͳҲָļ
> ```

Obviously, this is a problem due to an argument processing error somewhere inside the libraries that uses to porting the GNU toolchain to Windows (i.e. cygwin, msys, MinGW). Nothing can be fixed inside libmdbx to avoid this.
I can only recommend using CMake to build libmdbx on Windows (either don't use Windows at all).


> then I tried to build manually to see detail, which is as follows:
> ```
> PS C:\Users\asus\.nimble\pkgs\nimdbx-0.4.1\libmdbx-dist> gcc -std=gnu11 -O2 -g -Wall -Werror -Wextra -Wpedantic -ffunction-sections -fPIC -fvisibility=hidden -pthread -Wno-error=attributes  -DNDEBUG=1 -DMDBX_CONFIG_H="config.h" -ULIBMDBX_EXPORTS -c mdbx.c -o mdbx-static.o
> mdbx.c:16:10: error: #include expects "FILENAME" or <FILENAME>
>  #include MDBX_CONFIG_H
>           ^~~~~~~~~~~~~
> ...

@gogolxdong, you used a wrong quoting/escaping for the arguments.
Should be `'-DMDBX_CONFIG_H="config.h"'` instead of `-DMDBX_CONFIG_H="config.h"`.

**Comment by snej** [comment]
@gogolxdong -- You may be using an incompatible `make` tool. The errors reported in the first listing, like "`'case' is not recognized as an internal or external command`", come from the function at lines 49-55 of the GNUmakefile:
```
define uname2sosuffix
  case "$(UNAME)" in
    Darwin*|Mach*) echo dylib;;
...
```
I think your `make` tool is not compatible with GNU Make.

**Comment by gogolxdong** [comment]
cmake version 3.18.3
It doesn't show `#include expects "FILENAME" or <FILENAME>` with '-DMDBX_CONFIG_H="config.h"' ,but compile shows many errors with. And nimble test returns 
```shell
# Including library C:\nimdbx\libmdbx-dist\libmdbx.a
# Importing C:\nimdbx\libmdbx-dist\mdbx.h
# Saved to C:/Users/asus/nimcache/nimterop/toastCache/nimterop_3836952709.nim
C:\nimdbx\nimdbx\private\libmdbx.nim(56, 8) template/generic instantiation of `cImport` from 
here
C:\Users\asus\.nimble\pkgs\nimterop-0.6.13\nimterop\cimport.nim(704, 10) template/generic instantiation of `cImport` from here
C:\Users\asus\nimcache\nimterop\toastCache\nimterop_3836952709.nim(988, 48) Error: undeclared identifier: 'HANDLE'
       Tip: 17 messages have been suppressed, use --verbose to show them.
     Error: Execution failed with exit code 1
        ... Command: C:\Nim\bin\nim.exe c --noNimblePath -d:NimblePkgVersion=0.4.1 --path:C:\Users\asus\.nimble\pkgs\nimterop-0.6.13 --path:C:\Users\asus\.nimble\pkgs\regex-0.16.2 --path:C:\Users\asus\.nimble\pkgs\unicodedb-0.9.0 --path:C:\Users\asus\.nimble\pkgs\unicodeplus-0.8.0 --path:C:\Users\asus\.nimble\pkgs\unicodedb-0.9.0 --path:C:\Users\asus\.nimble\pkgs\segmentation-0.1.0 --path:C:\Users\asus\.nimble\pkgs\unicodedb-0.9.0 --path:C:\Users\asus\.nimble\pkgs\cligen-1.2.2 --hints:off -r --path:. C:\nimdbx\tests\test1_Data
```

**Comment by erthink** [comment]
I just added an amalgamated source code check during `make dist`, and these changes have passed CI-check for Linux, MacOS and FreeBSD.
This means that libmdbx's amalgamation script don't have any troubles, but work fine on Linux, FreeBSD and MacOS (and before was tested on Solaris, OpenSolaris, OpenIndiana, NetBSD, OpenBSD, DragonFly).

So I don't known any reason for these problems, other than a bugs in the Windows version of Unix utils (i.e. `make`, `sed`, `bash`, etc) or some drawback in the `nimdbx`'s build scripts. It is worth trying to use the latest versions of Cygwin, MSYS, MinGW and submit corresponding bug report(s).

I will be happy to help in anyway, but for now I do not know what can be fixed in `libmdbx` to avoid these problems with Windows.

**Comment by gogolxdong** [comment]
`Error: undeclared identifier: 'HANDLE'` is issue of nimterope, has nothing to do with libmdbx on windows. Thanks for replying!
